modules/stochastic_tools/src/mappings/PODMapping.C

```diff
@@ -0,0 +1,432 @@
+      "so the real number of modes might be lower than this. This is also used for setting the "
+      "subspace sizes for distributed singular value solves. By default the subspace used for the "
+      "SVD is twice as big as the number of requested vectors. For more information see the SLEPc "
+      "manual.");
+  params.addParam<std::vector<Real>>(
+      "energy_threshold",
+      std::vector<Real>(),
+      "The energy threshold for the automatic truncation of the number of modes. In general, the "
+      "lower this number is the more information is retained about the system by keeping more POD "
+      "modes.");
+  params.addParam<std::string>("extra_slepc_options",
+                               "",
+                               "Additional options for the singular/eigenvalue solvers in SLEPc.");
+  return params;
+}
+
+PODMapping::PODMapping(const InputParameters & parameters)
+  : MappingBase(parameters),
+    _num_modes(getParam<std::vector<dof_id_type>>("num_modes")),
+    _energy_threshold(getParam<std::vector<Real>>("energy_threshold")),
+    _left_basis_functions(
+        isParamValid("filename")
+            ? setModelData<std::map<VariableName, std::vector<DenseVector<Real>>>>(
+                  "left_basis_functions")
+            : declareModelData<std::map<VariableName, std::vector<DenseVector<Real>>>>(
+                  "left_basis_functions")),
+    _right_basis_functions(
+        isParamValid("filename")
+            ? setModelData<std::map<VariableName, std::vector<DenseVector<Real>>>>(
+                  "right_basis_functions")
+            : declareModelData<std::map<VariableName, std::vector<DenseVector<Real>>>>(
+                  "right_basis_functions")),
+    _singular_values(
+        isParamValid("filename")
+            ? setModelData<std::map<VariableName, std::vector<Real>>>("singular_values")
+            : declareModelData<std::map<VariableName, std::vector<Real>>>("singular_values")),
+    _extra_slepc_options(getParam<std::string>("extra_slepc_options"))
+{
+  if (!isParamValid("filename"))
+  {
+    if (_num_modes.size() != _variable_names.size())
+      paramError("num_modes", "The number of modes should be defined for each variable!");
+
+    for (const auto & mode : _num_modes)
+      if (!mode)
+        paramError("num_modes", "The number of modes should always be a positive integer!");
+
+    if (_energy_threshold.size())
+    {
+      if (_energy_threshold.size() != _variable_names.size())
+        paramError("energy_threshold",
+                   "The energy thresholds should be defined for each variable!");
+
+      for (const auto & threshold : _energy_threshold)
+        if (threshold < 0 || threshold >= 1)
+          paramError("energy_threshold",
+                     "The energy thresholds should always be in the [0,1) range!");
+    }
+
+#if PETSC_VERSION_LESS_THAN(3, 14, 0)
+    mooseError("PODMapping is not supported with PETSc version below 3.8!");
+#else
```

Comment:
When we tests with min gcc/clang we error out because even though it can find headers for slepc, it doesn't have the SVD routines. I'll check this out though!